lib/events.js

```diff
@@ -171,66 +174,67 @@ EventEmitter.prototype.emit = function emit(type, ...args) {
 };
 
 function _addListener(target, type, listener, prepend) {
-  var m;
-  var events;
-  var existing;
-
   if (typeof listener !== 'function') {
     const errors = lazyErrors();
     throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'listener', 'Function');
   }
 
-  events = target._events;
+  var events = target._events;
+  var list;
   if (events === undefined) {
     events = target._events = Object.create(null);
     target._eventsCount = 0;
   } else {
     // To avoid recursion in the case that type === "newListener"! Before
     // adding it to the listeners, first emit "newListener".
     if (events.newListener !== undefined) {
-      target.emit('newListener', type,
-                  listener.listener ? listener.listener : listener);
+      target.emit('newListener', type, listener.listener || listener);
 
       // Re-assign `events` because a newListener handler could have caused the
       // this._events to be assigned to a new object
       events = target._events;
     }
-    existing = events[type];
+    list = events[type];
   }
 
-  if (existing === undefined) {
+  if (list === undefined) {
     // Optimize the case of one listener. Don't need the extra array object.
-    existing = events[type] = listener;
+    events[type] = listener;
     ++target._eventsCount;
-  } else {
-    if (typeof existing === 'function') {
-      // Adding the second element, need to change to array.
-      existing = events[type] =
-        prepend ? [listener, existing] : [existing, listener];
-      // If we've already got an array, just append.
-    } else if (prepend) {
-      existing.unshift(listener);
-    } else {
-      existing.push(listener);
-    }
+    return target;
+  }
 
-    // Check for listener leak
-    if (!existing.warned) {
-      m = $getMaxListeners(target);
-      if (m && m > 0 && existing.length > m) {
-        existing.warned = true;
-        // No error code for this since it is a Warning
-        const w = new Error('Possible EventEmitter memory leak detected. ' +
-                            `${existing.length} ${String(type)} listeners ` +
-                            'added. Use emitter.setMaxListeners() to ' +
-                            'increase limit');
-        w.name = 'MaxListenersExceededWarning';
-        w.emitter = target;
-        w.type = type;
-        w.count = existing.length;
-        process.emitWarning(w);
-      }
+  if (typeof list === 'function') {
+    // Adding the second element, need to change to array.
+    list = events[type] = prepend ? [listener, list] : [list, listener];
+  } else if (list.emitting) {
+    const { warned } = list;
+    list = events[type] =
+      arrayCloneWithElement(list, listener, prepend ? 1 : 0);
+    if (warned) {
+      list.warned = true;
+      return target;
     }
+  } else if (prepend) {
+    list.unshift(listener);
+  } else {
+    list.push(listener);
+  }
+
+  // Check for listener leak
+  const m = $getMaxListeners(target);
```

Comment:
Did you benchmark this particular change? I would think that checking `list.warned` is less expensive than calling out to `$getMaxListeners()` every time, even after `.warned === true`?

Reply:
Accessing `.warned` or any other named key on an Array is really expensive from the testing I've done. Also, even if it wasn't, it's better to optimize for the case where `warned === false` since that's going to be more common.